Sources/_MatchingEngine/Regex/Parse/Mocking.swift

```diff
@@ -1,107 +1,199 @@
+  func load() -> UInt8? { load(offset: 0) }
+  func advance(_ n: Int = 1) {
+    precondition(current + n <= end, "Cannot advance past end")
+    current = current.advanced(by: n)
   }
 
-  guard eat() == ascii("'") else {
-    throw LexError.notRegexLiteralStart
+  func tryEat(_ utf8: String.UTF8View) -> Bool {
+    for (i, idx) in utf8.indices.enumerated() {
+      guard load(offset: i) == utf8[idx] else { return false }
+    }
+    advance(utf8.count)
+    return true
   }
 
-  let syntax: SyntaxOptions
-  let delimiter = eat()
-  switch delimiter {
-  case ascii("|"): syntax = .modern
-  case ascii("/"): syntax = .traditional
-  default: return nil
+  // Try to lex the opening delimiter.
+  guard let delimiter = Delimiter.allCases.first(
+    where: { tryEat($0.opening.utf8) }
+  ) else {
+    throw LexError(.unknownDelimiter, resumeAt: current.successor())
   }
 
   let contentsStart = current
-  let count: Int = try {
-    while true {
-      switch load() {
-      case 0: throw LexError.endOfString
-      case ascii("\\"):
-        // Skip next byte
-        advance()
-        advance()
-      case delimiter:
-        if load(offset: 1) == ascii("'") {
-          return current - contentsStart
-        }
+  while true {
+    switch load() {
+    case nil, ascii("\n"), ascii("\r"):
+      throw LexError(.endOfString, resumeAt: current)
+
+    case ascii("\\"):
+      // Skip next byte.
+      advance(2)
+
+    default:
+      // Try to lex the closing delimiter.
+      let contentsEnd = current
+      guard tryEat(delimiter.closing.utf8) else {
         advance()
+        continue
+      }
+
+      // Form a string from the contents and make sure it's valid UTF-8.
+      let count = contentsEnd - contentsStart
+      let contents = UnsafeRawBufferPointer(
+        start: contentsStart, count: count)
+      let s = String(decoding: contents, as: UTF8.self)
 
-      default: advance()
+      guard s.utf8.elementsEqual(contents) else {
+        throw LexError(.invalidUTF8, resumeAt: current)
       }
+      return (contents: s, delimiter, end: current)
     }
-  }()
+  }
+}
 
-  let contents = UnsafeRawBufferPointer(
-    start: contentsStart, count: count)
-  let s = String(decoding: contents, as: UTF8.self)
+private func copyCString(_ str: String) -> UnsafePointer<CChar>? {
+  let count = str.utf8.count + 1
+  return str.withCString {
+    assert($0[count-1] == 0)
+    let ptr = UnsafeMutablePointer<CChar>.allocate(capacity: count)
+    ptr.initialize(from: $0, count: count)
+    return UnsafePointer(ptr)
+  }
+}
 
-  guard s.utf8.elementsEqual(contents) else {
-    throw LexError.invalidUTF8
+/// Public interface for libswift.
+///
+/// Lex a regular expression literal starting at `inputPtr`, making sure not to
+/// lex past `bufferEndPtr`. The pointer at which to resume lexing is returned,
+/// or nil if this is not a regex literal. The `errOut` parameter will be set
+/// if an error is encountered.
+public func libswiftLexRegexLiteral(
```

Comment:
Can this be `@usableFromInline internal` instead?

Reply:
Ah good point! Updated